lib/phoenix/view.ex

```diff
@@ -241,23 +240,31 @@ defmodule Phoenix.View do
     |> render_within(module, template)
   end
 
-  defp render_within({{layout_mod, layout_tpl}, assigns}, inner_mod, inner_tpl) do
-    assigns = Map.merge(assigns, %{view_module: inner_mod,
-                                   view_template: inner_tpl})
-
-    render_layout(layout_mod, layout_tpl, assigns)
-  end
-
   defp render_within({false, assigns}, module, template) do
-    assigns = Map.merge(assigns, %{view_module: module,
-                                   view_template: template})
+    assigns = Map.merge(assigns, %{view_module: module, view_template: template})
     module.render(template, assigns)
   end
 
-  defp render_layout(layout_mod, layout_tpl, assigns) do
+  defp render_within({layout, assigns}, inner_mod, inner_tpl) do
+    assigns = Map.merge(assigns, %{view_module: inner_mod, view_template: inner_tpl})
+    render_layout(layout, assigns)
+  end
+
+  defp render_layout({layout_mod, layout_tpl}, assigns)
+       when is_atom(layout_mod) and is_binary(layout_tpl) do
```

Comment:
Could additionally check here that the layout model is valid by checking
 `:erlang.function_exported(layout_mod, :render, 2)`

Reply:
I think it's good enough to simply match on `{:layout_mod, tpl}` as you've done